### ISSUE TITLE
Lists as category values

### ISSUE THREAD

**Issue by mimischi** [datreant/datreant#164]
I've stumbled upon a few cases where I'd wanted to set a list as the value for a category. A MD specific example: I run a bunch of simulations that differ in their membrane composition. Now I'd like to have a category `lipids` that holds a list with lipid names i.e.:

```python
{
    "categories": {
        "lipids": ["POPC", "DOPC", "DOPA"]
    }
}
```

A current work around would be put these as a string and have some delimiter (`,`) to handle such cases.

Is there any specific reason why `datreant` does not support lists as keys to categories?

**Comment by richardjgowers** [comment]
It might be that lists aren’t hashable (because they’re mutable). If you
try with a tuple does it work?

**Comment by mimischi** [comment]
Category values are currently limited to `int`, `float`, `str` and `bool`.

https://github.com/datreant/datreant/blob/dcd432fe991ec39c871aeb0f87693f1ff7c2c7cc/src/datreant/metadata.py#L455-L465

**Comment by richardjgowers** [comment]
Ah ok, yeah tuple values (or any json convertible?) seem like they should
work... hack it in and see if anything breaks?

**Comment by dotsdl** [comment]
Hey @mimischi, thanks for the thought you've put into this! I have a question, however: what should happen for `Bundle.groupby`? This places values of categories as keys in a dictionary, which won't work for lists. This is where the hashable requirement comes from for category values.

We will need to change the behavior of `Bundle.groupby` to accommodate this change. Perhaps instead of lists, we accept tuples as values (so long as the values in the tuple are JSON-serializble and hashable)?

**Comment by mimischi** [comment]
Good point. Function-wise I would expect `Bundle.groupby` to return bundles with the same list items, i.e., in my example with the same lipids in the list. Although will order matter here?

Accepting tuples instead of lists wouldn't really change anything for the user, would it? One could just cast the list into a tuple in the background.

**Comment by richardjgowers** [comment]
Maybe it's safer to use hashable&immutable items (ie tuple not list here) for values throughout.  So silently convert lists to tuples when setting/getting?

**Comment by dotsdl** [comment]
We could allow tuples as category values, but not dicts or lists. Internally in JSON these tuples will be written out as lists, but read back in as tuples. Does this satisfy your needs @mimischi? I think we could make this work, so long as the tuple contents are hashable (read: immutable).

**Comment by mimischi** [comment]
Yes, setting any sequence (tuple or list) would help!

**Comment by kain88-de** [comment]
Are you sure setting a tuple is in your interest here? Wouldn't tags be more suited for you? The tuple value will be *order dependent*, so having the right order in a groupby matters. Everytime you add this key you need the same value in the exact same order! You won't be able to search for simulations that have any 'DOPC' lipids.  You can even leverage fuzzyfinder to search for `DOPV` and will get all similar tags.

**Comment by richardjgowers** [comment]
Ahhh order is a good point. So surely a set is the way to define multiple
values (if we’re doing this)

**Comment by mimischi** [comment]
Tags would allow to filter by specific lipids, but would require you to know _which_ lipids are used in a given system. This is the information I want to off-load onto categories in datreant.

**Comment by dotsdl** [comment]
Leaving this for later, since I just had an interesting discussion with @vyasr and @csadorf. I think we could expand the acceptable values for categories to "strings, numbers, (floats, ints), booleans, or sets of any of the above", with nesting of sets allowed. However, this then still creates issues for groupby, although we might be able to get around them with sorting. Need to think on this one more.

**Comment by richardjgowers** [comment]
Recursive hash?
  

  
  

  
  
>
